ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topk/TopkKudaf.java

```diff
@@ -26,35 +29,33 @@
 import java.util.List;
 import java.util.Map;
 
-import io.confluent.ksql.function.KsqlAggregateFunction;
-import io.confluent.ksql.parser.tree.Expression;
-import io.confluent.ksql.util.ArrayUtil;
-import io.confluent.ksql.util.KsqlException;
-
-
 public class TopkKudaf<T> extends KsqlAggregateFunction<T, T[]> {
-
-  private Integer tkVal;
-  private T[] tempTopkArray;
-  private Class<T> ttClass;
+  private final int topKSize;
+  private final T[] tempTopKArray;
+  private final Class<T> clazz;
+  private final Schema initialValue;
+  private final List<Schema> arguments;
 
   TopkKudaf(int argIndexInValue,
-            Integer tkVal,
-            Class<T> ttClass) {
+            int topKSize,
+            Schema initialValue,
+            List<Schema> arguments,
+            Class<T> clazz) {
     super(argIndexInValue,
-          (T[]) Array.newInstance(ttClass,tkVal),
```

Comment:
`Array.newInstance` is only able to allocate array at most 255 size, which is quite a small limit for streaming application.

Reply:
@satybald `Array.newInstance` can create array with bigger than 255 size. The 255 limit is for the number of dimensions in the array not the size of each dimension. Here we have one dimension only.

Reply:
@hjafarpour my bad, you're right. Array.newInstance can create array up to Int.MAX_VALUE for one dimension. Thought, `Array.newInstance` uses reflection to create a new instance. The rule of thumb is to avoid reflection calls as it incurs performance overhead.


> Reflection is powerful, but should not be used indiscriminately. If it is possible to perform an operation without using reflection, then it is preferable to avoid using it. The following concerns should be kept in mind when accessing code via reflection.

[https://docs.oracle.com/javase/tutorial/reflect/index.html](https://docs.oracle.com/javase/tutorial/reflect/index.html)

Reply:
@hjafarpour good catch. I've removed it.